core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -768,11 +768,13 @@ private[deploy] class Worker(
     }
   }
 
-  private[deploy] def decommissionSelf(): Boolean = {
+  private[deploy] def decommissionSelf(fromMaster: Boolean): Boolean = {
     if (conf.get(config.DECOMMISSION_ENABLED)) {
       logDebug("Decommissioning self")
       decommissioned = true
-      sendToMaster(WorkerDecommission(workerId, self))
+      if (!fromMaster) {
+        sendToMaster(WorkerDecommissioned(workerId))
```

Comment:
I wonder if it is trite to add a comment like: "No need to notify the master if the decommission message already came from it"